rmf_task/src/rmf_task/BackupFileManager.cpp

```diff
@@ -214,8 +240,58 @@ BackupFileManager::Group::Group()
 //==============================================================================
 std::optional<std::string> BackupFileManager::Robot::read() const
 {
-  throw std::runtime_error(
-    "[BackupFileManager::Robot::read] not implemented yet");
+  if (!std::filesystem::exists(_pimpl->robot_directory))
+    throw std::runtime_error("[BackupFileManager::Robot::read] Directory " + 
+      _pimpl->robot_directory.string() + " missing. This should not happen.");
+
+  if (std::filesystem::is_empty(_pimpl->robot_directory))
+    return std::nullopt;
+
+  // Check for foreign files
+  auto directory_it = std::filesystem::directory_iterator(_pimpl->robot_directory);
+  for(auto& p: directory_it)
+  {
+    auto filename = p.path().filename().string();
+    if (filename.compare(_pimpl->backup_file_name) != 0 && filename.compare(_pimpl->pre_backup_file_name) != 0)
+      throw std::runtime_error("[BackupFileManager::Robot::read] Foreign file " + 
+        filename + " found. This should be removed.");
+  }
+
+  // At this point, file is either backup_file_name, or .backup_file_name, or both
+  const std::string pre_backup_file_path = _pimpl->robot_directory.string() + "/" + _pimpl->pre_backup_file_name;
+  const std::string backup_file_path = _pimpl->robot_directory.string() + "/" + _pimpl->backup_file_name;
+
+  if (std::filesystem::exists(backup_file_path))
+  {
+    if (std::filesystem::exists(pre_backup_file_path))
+    {
+      //suspicious to have both backup files, something definitely broke in the previous run.
+      std::cout << "[BackupFileManager::Robot::read] Multiple backup files found. This suggests an error with the previous \
+        backup run. Using the older edited backup file.." << std::endl;
+      std::filesystem::remove(pre_backup_file_path);
+    }
+
+    std::ifstream backup;
+    backup.open(backup_file_path, std::ios::in);
```

Comment:
I'm somehow unable to get this to work, when I try this the file never gets created ( although no error gets thrown). I probably am missing something, since the docs do seem to suggest this is a valid approach.

Reply:
It's very suspicious that it would stop working :thinking: 

It's tempting to ignore this and just accept the code as-is, but I'm too curious about why the behavior wouldn't match the documentation, so I might take a closer look at this.